feat(git-node): add release promotion step

aduh95 opened the pull request:
You can try it with `git node release -S --promote https://github.com/nodejs/node/pull/53945`.

Supersedes https://github.com/nodejs/node-core-utils/pull/402

/cc @nodejs/releasers

RafaelGSS commented on lib/promote_release.js:
```diff
@@ -0,0 +1,437 @@
+
+    cli.separator();
+    cli.ok(`Release promotion for ${version} complete.\n`);
+    cli.info(
```
Maybe add a TODO to automate all these steps too?

aduh95 replied:
I think the only step that can be automated is "Create the release on GitHub" (although simply printing out which `gh` command to use might be preferable, it might be a bad idea to have NCU token with permission to create releases), I'll see if I can add it – the "Create the blog post for nodejs.org" step might be automatable as well, but a cross-repo script call would feel quite brittle.

aduh95 replied:
https://github.com/nodejs/node-core-utils/pull/835/commits/66fdab573af0456ae0f487597dee0f9a0d9f0aa2

RafaelGSS commented on lib/promote_release.js:
```diff
@@ -0,0 +1,437 @@
+      cli.warn(`Aborting release promotion for version ${version}`);
+      throw new Error('Aborted');
+    }
+    await this.cherryPickToDefaultBranch();
```
Shouldn't it be encapsulated in a try/catch? Also, we might want to update the session status to --continue after an error

aduh95 replied:
A failed cherry-pick does not throw, so the try/catch seems unnecessary. I agree the `--continue` would be nice, but that does feel more involved than I'd like, maybe it can be a follow up.

RafaelGSS replied:
Alright

marco-ippolito reviewed (approved):
LGTM I tested it with dry run for v20.16.0 (I stopped at tagging) but so far its great

RafaelGSS commented on components/git/release.js:
```diff
@@ -1,14 +1,24 @@
+const RELEASERS = 'releasers';
 
 const releaseOptions = {
+  dryRun: {
```
It should be enabled by default

aduh95 replied:
I'm fine with having `--dry-run` being the default, I agree it somewhat makes sense; does anyone have a suggestion for a good name for such option? I assume `--no-dry-run` is not very self explanatory.

targos replied:
`--run` ?

aduh95 replied:
Went with `--run` in https://github.com/nodejs/node-core-utils/pull/835/commits/5ccb47472540e82d2da38fa79fd750ae9804b92d

RafaelGSS reviewed (changes requested):
Tested with today's release and:

* Wrong warning:
```
   ⚠  The release date (2024-08-06) does not match the system date for today (2024-08-06).
```
* Github CI is green, but it says failing
```
? Do you want to proceed anyway? Yes
✔  Jenkins CI is passing
✘  GitHub CI is failing for #54123
```
* I have resolved the conflicts on `main`, but I though the script would do the `git cherry-pick --continue` for me, so it failed. Maybe we should clarify that in the prompt:
```
? Finished resolving cherry-pick conflicts? Yes
   ✘  Invalid Release commit message: deps: update acorn to 8.12.1
Error: Aborted
    at ReleasePromotion.validateReleaseCommit (file:///home/rafaelgss/repos/os/node-core-utils/lib/promote_release.js:221:13)
    at ReleasePromotion.promote (file:///home/rafaelgss/repos/os/node-core-utils/lib/promote_release.js:155:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```
* It doesn't store state, so after the above error I couldn't start from where I stopped.
* It could be a bit more clear that you should run the commands in a separate terminal, for instance:
```
? Tag and sign the release? Yes
⠸ Setting up for next release[v22.6.0-proposal a1e60bc6df] Working on v22.6.1
 1 file changed, 2 insertions(+), 2 deletions(-)
✔  Successfully set up for next release
   ℹ  Run the following commands to merge the staging branch:
   ℹ  git push upstream HEAD:refs/heads/v22.x
   ℹ  git push upstream HEAD:refs/heads/v22.x-staging
```
* When an error happens, we should at least print what are the remaining steps so users can do it manually

Overall, amazing work!

Full log

<details>

```console
➜  node (v22.6.0-proposal) ../node-core-utils/bin/git-node.js release -S --dryRun --promote https://github.com/nodejs/node/pull/54123
✔  Received member information of nodejs/releasers
✔  RafaelGSS is a Releaser
✔  Done loading data for nodejs/node/pull/54123
   ℹ  Last V8 CI on 2024-08-05T17:31:12Z: https://ci.nodejs.org/job/node-test-commit-v8-linux/6109/
   ℹ  Last Full PR CI on 2024-08-05T17:31:12Z: https://ci.nodejs.org/job/node-test-pull-request/60889/
   ℹ  Last CITGM CI on 2024-08-05T17:31:12Z: https://ci.nodejs.org/job/citgm-smoker/3469/
⠋ Querying data for job/node-test-pull-request/6109/   ✘  GitHub CI is still running
   ℹ  This PR was created on Tue, 30 Jul 2024 13:13:18 GMT
   ✔  Approvals: 3
   ✔  - Matteo Collina (@mcollina) (TSC): https://github.com/nodejs/node/pull/54123#pullrequestreview-2207716135
   ✔  - Antoine du Hamel (@aduh95) (TSC): https://github.com/nodejs/node/pull/54123#pullrequestreview-2207729636
   ✔  - Marco Ippolito (@marco-ippolito) (TSC): https://github.com/nodejs/node/pull/54123#pullrequestreview-2221494930
   ✔  Last Jenkins CI successful
   ⚠  The release date (2024-08-06) does not match the system date for today (2024-08-06).
--------------------------------------------------------------------------------
? Do you want to proceed anyway? Yes
✔  Jenkins CI is passing
✘  GitHub CI is failing for #54123
--------------------------------------------------------------------------------
? Do you want to proceed? Yes
✔  #54123 has necessary approvals
--------------------------------------------------------------------------------
? Tag and sign the release? Yes
⠸ Setting up for next release[v22.6.0-proposal a1e60bc6df] Working on v22.6.1
 1 file changed, 2 insertions(+), 2 deletions(-)
✔  Successfully set up for next release
   ℹ  Run the following commands to merge the staging branch:
   ℹ  git push upstream HEAD:refs/heads/v22.x
   ℹ  git push upstream HEAD:refs/heads/v22.x-staging
--------------------------------------------------------------------------------
? Ready to continue? Yes
--------------------------------------------------------------------------------
? Cherry-pick release commit to the default branch? Yes
Switched to branch 'main'
   ✔  No git cherry-pick in progress
   ✔  No git am in progress
   ✔  No git rebase in progress
--------------------------------------------------------------------------------
? Do you want to try reset the local main branch to upstream/main? Yes
⠇ Bringing upstream/main up to date...remote: Enumerating objects: 193, done.
remote: Counting objects: 100% (181/181), done.
remote: Compressing objects: 100% (95/95), done.
remote: Total 193 (delta 110), reused 124 (delta 82), pack-reused 12
Receiving objects: 100% (193/193), 339.12 KiB | 1.71 MiB/s, done.
Resolving deltas: 100% (116/116), completed with 45 local objects.116)
From github.com:nodejs/node
 * branch                  main       -> FETCH_HEAD
   c6aeddf140..54119757a3  main       -> upstream/main
✔  upstream/main is now up-to-date
main is out of sync with upstream/main. Mismatched commits:
 - 269175b2b4 doc: update security-release process to automated one
 - be7a0c5780 module: do not warn for typeless package.json when there isn't one
 - c40c41c285 src: expose LookupAndCompile with parameters
 - cf9a8140aa module: fix extensionless typescript in cjs loader
 - d7941b24ae test: do not swallow uncaughtException errors in exit code tests
 - 3c50297e9e typings: apply lint
 - 890760b8e5 console: fix issues with frozen intrinsics
 - 027710e408 doc: move numCPUs require to top of file in cluster CJS example
 - 7485ad817a crypto: include NODE_EXTRA_CA_CERTS in all secure contexts by default
 - a5decd4c8d build: drop experimental support for Windows <10
 - 8f9c660d92 deps: update corepack to 0.29.3
 - 4d1cc7618e tools: update eslint to 9.8.0
 - f9f9a421a8 path: use the correct name in `validateString`
 - e7edcf38cd buffer: use faster integer argument check
 - b4fd1fd63f url: improve resolveObject with ObjectAssign
 - f212674452 doc: clarify `useCodeCache` setting for cross-platform SEA generation
 - 49a9ba43cf doc: expand troubleshooting section
 - 4d1d88118b meta: add typescript team to codeowners
 - 67bc6a4a3e src: account for OpenSSL unexpected version
 - 7941b4b333 doc: remove unused imports from worker_threads.md
 - 1fd170a7fc stream: throw TypeError when criteria fulfilled in getIterator
 - e2b7e41e23 child_process: remove unused internal event
 - 00c0644747 module: remove outdated comment
 - 8e1e3a8dea typings: add missing binding function `writeFileUtf8()`
 - 76c8ba9d9b buffer: use native copy impl
 - 2f7a354f14 doc: update list of Triagers on the `README.md`
 - 1e3a3470a5 build: add `--without-amaro` build flag
 - 13ca021da3 doc: fix worker threadId/destination typo
 - 0c1877a82a test: add coverage for webstorage quota
 - d1229eeca4 lib: rewrite AsyncLocalStorage without async_hooks
 - 492032f34c doc: add missing new lines to custom test reporter examples
 - 358ff748ea lib,permission: support Buffer to permission.has
 - acbd0be0da buffer: faster type check
 - 5d6c76adee module: fix strip-types interaction with detect-module
 - d172da8d01 test: add initial pull delay and prototype pollution prevention tests
 - bd42e4c6a7 test: fix typo in test
 - ea5e77b1a5 benchmark: use assert.ok instead of assert
 - 76d10a10fb doc: add documentation for blob.bytes() method
 - 67f713731c test_runner: run after hooks even if test is aborted
 - c6cf313bca doc: refresh instructions for building node from source
 - a3cd290f24 buffer: add JSDoc to blob bytes method
 - e6ee6e2901 lib: fix typos in comments within internal/streams
 - ca2ed88f94 test_runner: remove outdated comment
 - b3013e120a tools: add strip-types to label system
 - a8166880f1 watch: reload changes in contents of --env-file
 - fafc845089 stream: make checking pendingcb on WritableStream backward compatible
 - 3616ace637 inspector: provide detailed info to fix DevTools frontend errors
 - 046343ea9d build: warn for GCC versions earlier than 12.2
 - cf21220893 src: move spkac methods to ncrypto
 - bc6f895030 doc: move release key for Myles Borins
 - ec7ee6c95e deps: update amaro to 0.0.5
 - b95eac5061 deps: update amaro to 0.0.6
 - 981b5e3279 doc, meta: replace command with link to keys
 - c6aeddf140 sea: don't set code cache flags when snapshot is used
 - 926503b669 src: shift more crypto impl details to ncrypto
 - a5200da1d0 doc: move `onread` option from `socket.connect()` to `new net.socket()`
 - 28adfbe746 meta: bump `ossf/scorecard-action` from 2.3.3 to 2.4.0
 - e55eb45d34 meta: bump `github/codeql-action` from 3.25.11 to 3.25.15
 - 6edbc04a3b meta: bump `actions/setup-node` from 4.0.2 to 4.0.3
 - 552df3edab meta: bump `step-security/harden-runner` from 2.8.1 to 2.9.0
 - 185218887b meta: bump actions/setup-python from 5.1.0 to 5.1.1
 - 66a34da003 meta: bump `actions/download-artifact` from 4.1.7 to 4.1.8
 - 5b32511f54 meta: bump `actions/upload-artifact` from 4.3.3 to 4.3.4
 - f182c6c0bc deps: update simdutf to 5.3.1
 - 5f52c8ae54 deps: update undici to 6.19.5
 - accf2acdfb build: support `lint-js-fix` in `vcbuild.bat`
 - 1bcdba233d test: update wpt test for streams
 - 2b41c237c0 doc: add links to security steward companies
 - 1f69085db5 test: set test-structuredclone-jstransferable non-flaky
 - 24ed890663 doc: fix sea assets example
 - 5bfebfec88 test_runner: make mock_loader not confuse CJS and ESM resolution
 - c852e222cb path: fix relative on Windows
 - fd49758796 doc: add esm examples to node:console
 - 277ed9fd86 doc: add KevinEady as a triager
 - b47aa708a8 tools: add workflow to ensure `README` lists are in sync with gh teams
 - e1e7a417da benchmark: remove force option as force defaults to true
 - 359b4284bc test: add subtests to test-node-run
 - 54119757a3 deps: update acorn to 8.12.1
--------------------------------------------------------------------------------
? Reset to upstream/main? Yes
HEAD is now at 54119757a3 deps: update acorn to 8.12.1
   ✔  Reset to upstream/main
Auto-merging CHANGELOG.md
Auto-merging doc/api/cli.md
Auto-merging doc/api/errors.md
Auto-merging doc/api/stream.md
Auto-merging doc/api/test.md
CONFLICT (content): Merge conflict in doc/api/test.md
Auto-merging src/node_version.h
CONFLICT (content): Merge conflict in src/node_version.h
error: could not apply b096e41b48... 2024-08-06, Version 22.6.0 (Current)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".
--------------------------------------------------------------------------------
   ℹ  Resolve the conflicts and commit the result
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
? Finished resolving cherry-pick conflicts? Yes
   ✘  Invalid Release commit message: deps: update acorn to 8.12.1
Error: Aborted
    at ReleasePromotion.validateReleaseCommit (file:///home/rafaelgss/repos/os/node-core-utils/lib/promote_release.js:221:13)
    at ReleasePromotion.promote (file:///home/rafaelgss/repos/os/node-core-utils/lib/promote_release.js:155:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```
</details>

aduh95 commented:
> * Github CI is green, but it says failing
> 
> ```
> ? Do you want to proceed anyway? Yes
> ✔  Jenkins CI is passing
> ✘  GitHub CI is failing for #54123
> ```

You probably already know that, but in case someone else doesn't know: that's not an issue introduced by this PR, it's a recurring issue that affects all PRs open from nodejs/node (sometimes it says "CI is failing", sometimes it says "CI is still running")

aduh95 commented on lib/promote_release.js:
```diff
@@ -0,0 +1,459 @@
+    }
+    data.date = releaseCommitMessage.slice(0, 10);
+    const systemDate = new Date().toISOString().slice(0, 10);
+    if (this.date !== systemDate) {
```
```suggestion
    if (data.date !== systemDate) {
```

aduh95 commented:
> * When an error happens, we should at least print what are the remaining steps so users can do it manually

That seems quite hard to do, because either it's a error case we can predict (and in this case, we should have the code account for it, not hard fail), either it's something we can't predict (and in which case, the steps to be taken are also hard to predict).

> * I though the script would do the `git cherry-pick --continue` for me, so it failed. Maybe we should clarify that in the prompt

We can have the script check if `.git/CHERRY_PICK_HEAD` exists, and run `git cherry-pick --continue` at this point. I'll push something for that

aduh95 commented:
> * It doesn't store state, so after the above error I couldn't start from where I stopped.

Hopefully that's not a blocking concern, as I don't think I'll be willing in implementing that in this PR.



> * It could be a bit more clear that you should run the commands in a separate terminal

Do you have a suggestion? Is this a blocking concern?

RafaelGSS reviewed (approved):
> That seems quite hard to do, because either it's a error case we can predict (and in this case, we should have the code account for it, not hard fail), either it's something we can't predict (and in which case, the steps to be taken are also hard to predict).

I meant for errors like:

```
Invalid Release commit message: deps: update acorn to 8.12.1
```

But, I think I can work on that in a follow-up PR.

> Hopefully that's not a blocking concern, as I don't think I'll be willing in implementing that in this PR.

Surely not.

> Do you have a suggestion? Is this a blocking concern?

No, I can handle it later.

ruyadorno commented:
exciting 🤩

targos commented:
Thanks for working on this! I will try it the next time I make a release.

RafaelGSS commented on lib/promote_release.js:
```diff
@@ -0,0 +1,469 @@
+      '/^<a id="[0-9]+\\.[0-9]+\\.[0-9]+"><\\x2fa>$/{' +
+      'print buf; if(firstLine == "") firstLine = $0; else buf = $0' +
+    `}' doc/changelogs/CHANGELOG_V${
+      this.versionComponents.major}.md | gh release create ${this.version} --verify-tag --latest${
```
```suggestion
      this.versionComponents.major}.md | gh release create v${this.version} --verify-tag --latest${
```

RafaelGSS commented:
What's missing to have it landed?

aduh95 commented:
@targos said he wanted to give it a shot, so I'm fine waiting on his input – but also fine landing it if folks think it's ready to land.

RafaelGSS commented:
So, I gave a shot today for v22.9.0 and something got wrong. It conflicts (as expected) but on v22.x-staging branch instead of `main` branch. I think you merge with main got wrong

```
? Do you want to try reset the local v22.x-staging branch to upstream/v22.x-staging? Yes
⠦ Bringing upstream/v22.x-staging up to date...From github.com:nodejs/node
 * branch                    v22.x-staging -> FETCH_HEAD
✔  upstream/v22.x-staging is now up-to-date
v22.x-staging is out of sync with upstream/v22.x-staging. Mismatched commits:
 - 4631be0311f 2024-09-17, Version 22.9.0 (Current)
 - 8a3482ec5e4 Working on v22.9.1
--------------------------------------------------------------------------------
? Reset to upstream/v22.x-staging? Yes
HEAD is now at 8a3482ec5e4 Working on v22.9.1
   ✔  Reset to upstream/v22.x-staging
Auto-merging src/node_version.h
CONFLICT (content): Merge conflict in src/node_version.h
error: could not apply 4631be0311f... 2024-09-17, Version 22.9.0 (Current)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".
--------------------------------------------------------------------------------
   ℹ  Resolve the conflicts and commit the result
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
? Finished resolving cherry-pick conflicts? Yes
   ℹ  You are running in dry-run mode, meaning NCU will not run the `git push` commands, you would need to copy-paste the following command in another terminal window. Alternatively, pass `--run` flag to ask NCU to run the command for you (might not work if you need to type a passphrase to push to the remote).
   ℹ  Run the following commands to push to remote:
   ℹ  git push upstream v22.x-staging v22.9.0
```

RafaelGSS commented:
So it created 2 release commits, 

* 1 - Release commit
* 2 - Working on v22.9.1
* 3 - Release commit

aduh95 commented:
It expects `ncu-config get branch` to return `main`, could it be that it is set to `v22.x-staging` on your local machine?

RafaelGSS commented:
> It expects `ncu-config get branch` to return `main`, could it be that it is set to `v22.x-staging` on your local machine?

It was null

```
➜  nodejs-internal-benchmark git:(main) ✗ ncu-config get branch
undefined
```

aduh95 commented:
Oh! My bad, I didn't test that case, tbh I thought `main` would be the default already.

aduh95 commented:
So I tried reproducing by removing the setting for the branch name, and I immediately got an error:

```console
$ git node release -S --promote https://github.com/nodejs/node/pull/54966
   ⚠  You have not told git-node what branch you are trying to land commits on.
--------------------------------------------------------------------------------
   ℹ  For example, if your want to land commits on the `main` branch, you can run:

  $ ncu-config set branch main
--------------------------------------------------------------------------------
Error: Failed to create new session
    at new Session (file:///…/node-core-utils/lib/session.js:27:15)
    at new ReleasePromotion (file:///…/node-core-utils/lib/promote_release.js:22:5)
    at main (file:///…/node-core-utils/components/git/release.js:127:21)
…
```

Did you modify that check locally maybe? You should not be able to proceed with an empty value.

RafaelGSS commented:
I did not. I have just `gh pr checkout 835` and run the script. 

Actually, I found why. I normally use a local ncu configuration and it was set to v22.x-staging indeed. Can we automatically change it to `main` or at least exit the script if it wasn't set properly?

targos commented:
So, I tried the command for v20.18.0. Unfortunately, I didn't go far. At some point it tried to update my local staging branch from upstream but it was already pushed in an earlier step (the manual one, since I ran in dry-run mode by mistake, but the dry-run had already created the tag so it wasn't completely dry).
I guess I should have answered "no" to the questions about v20.x-staging but I was afraid that it would cancel the entire run.

<details>

````
$ node ../node-core-utils/bin/git-node.js release --promote https://github.com/nodejs/node/pull/55170
✔  Received member information of nodejs/releasers
✔  targos is a Releaser
✔  Done loading data for nodejs/node/pull/55170
   ℹ  Last Full PR CI on 2024-10-02T14:11:11Z: https://ci.nodejs.org/job/node-test-pull-request/62893/
   ℹ  Last CITGM CI on 2024-10-02T14:11:11Z: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/3484/
⠋ Querying data for job/node-test-pull-request/62893/   ✘  Last GitHub CI failed
   ℹ  This PR was created on Mon, 30 Sep 2024 08:59:29 GMT
   ✔  Approvals: 4
   ✔  - Matteo Collina (@mcollina) (TSC): https://github.com/nodejs/node/pull/55170#pullrequestreview-2342018931
   ✔  - Antoine du Hamel (@aduh95) (TSC): https://github.com/nodejs/node/pull/55170#pullrequestreview-2343657161
   ✔  - Richard Lau (@richardlau) (TSC): https://github.com/nodejs/node/pull/55170#pullrequestreview-2345406712
   ✔  - Ruy Adorno (@ruyadorno) (TSC): https://github.com/nodejs/node/pull/55170#pullrequestreview-2345570558
   ✔  Last Jenkins CI successful
✔  Jenkins CI is passing
✘  GitHub CI is failing for #55170
--------------------------------------------------------------------------------
? Do you want to proceed? Yes
✔  #55170 has necessary approvals
--------------------------------------------------------------------------------
? Tag and sign the release? Yes
⠼ Setting up for next release[v20.18.0-proposal 985262a4a82] Working on v20.18.1
 1 file changed, 2 insertions(+), 2 deletions(-)
✔  Successfully set up for next release
   ℹ  You are running in dry-run mode, meaning NCU will not run the `git push` commands, you would need to copy-paste the following command in another terminal window. Alternatively, pass `--run` flag to ask NCU to run the command for you (might not work if you need to type a passphrase to push to the remote).
   ℹ  Run the following commands to merge the staging branch:
   ℹ  git push upstream HEAD:refs/heads/v20.x HEAD:refs/heads/v20.x-staging
--------------------------------------------------------------------------------
? Ready to continue? Yes
--------------------------------------------------------------------------------
? Cherry-pick release commit to the default branch? Yes
Switched to branch 'v20.x-staging'
Your branch is behind 'upstream/v20.x-staging' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
   ✔  No git cherry-pick in progress
   ✔  No git am in progress
   ✔  No git rebase in progress
--------------------------------------------------------------------------------
? Do you want to try reset the local v20.x-staging branch to upstream/v20.x-staging? Yes
⠙ Bringing upstream/v20.x-staging up to date...From github.com:nodejs/node
 * branch                    v20.x-staging -> FETCH_HEAD
✔  upstream/v20.x-staging is now up-to-date
v20.x-staging is out of sync with upstream/v20.x-staging. Mismatched commits:
 - 7eebd17fa2c 2024-10-03, Version 20.18.0 'Iron' (LTS)
 - 985262a4a82 Working on v20.18.1
--------------------------------------------------------------------------------
? Reset to upstream/v20.x-staging? Yes
HEAD is now at 985262a4a82 Working on v20.18.1
   ✔  Reset to upstream/v20.x-staging
Auto-merging src/node_version.h
CONFLICT (content): Merge conflict in src/node_version.h
error: could not apply 7eebd17fa2c... 2024-10-03, Version 20.18.0 'Iron' (LTS)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".
hint: Disable this message with "git config advice.mergeConflict false"
--------------------------------------------------------------------------------
   ℹ  Resolve the conflicts and commit the result
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
? Finished resolving cherry-pick conflicts? Yes
   ✘  Invalid Release commit message: Working on v20.18.1
Error: Aborted
    at ReleasePromotion.validateReleaseCommit (file:///Users/mzasso/git/nodejs/node-core-utils/lib/promote_release.js:234:13)
    at ReleasePromotion.promote (file:///Users/mzasso/git/nodejs/node-core-utils/lib/promote_release.js:168:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
````

<details>

aduh95 commented:
FWIW you can delete the local tag as long as you haven't pushed it.

> ? Cherry-pick release commit to the default branch? Yes
> Switched to branch 'v20.x-staging'

It looks like it thinks the default branch is `v20.x-staging` and not `main`, which is unlikely to produce good results. What does `ncu-config get branch` return?

targos commented:
It is v20.x-staging, which is normal from my pov. I work on a v20.x release in my v20.x workspace

aduh95 commented:
> It is v20.x-staging, which is normal from my pov

Does it need to be `v20.x-staging` for something else? Landing backport PR maybe? Asking because I'm wondering if we should rename it to `defaultBranch` to avoid the confusion – or if I choose just make an API call to get the name of the default branch (and assume there's a local branch with that name)

targos commented:
I believe it's for backport PRs, but I'm not sure.

aduh95 commented:
I've confirmed that to land a (backport or not) PR to a staging branch, NCU request you to set a different branch in your config. I've updated the promotion script to make an API call to get the name of the default branch instead of relying on the config.
I've also added a fallback when the version tag already to not crash if it points to the right commit and is correctly signed.

aduh95 commented:
I've tried it with the v18.20.5 release, and did not get any blocker. I'm tempted to land it, we might still want to iterate on it but I think it's at a point where we can rely on it. @nodejs/releasers, wdyt?
